File: src/neutral_physics_engine/octree.py
"""
octree.py

This module implements the Barnes-Hut octree data structure for high-performance
N-body gravitational simulations. It provides an O(N log N) approximation algorithm
for computing accelerations and gravitational potentials, making it suitable for
large-scale astrophysical, planetary, or spacecraft dynamics simulations (NASA/ESA-grade
professional physics engines).

The Octree class builds a hierarchical 3D spatial partitioning tree where each node
stores the total mass and center-of-mass of all bodies it contains. This enables
the Barnes-Hut opening-angle criterion (controlled by `theta`) to decide whether to
approximate a distant cluster of bodies as a single point mass or to recurse into
children for higher accuracy.

Key features:
- Efficient tree construction and insertion with automatic subdivision
- Recursive force/acceleration evaluation with Barnes-Hut approximation
- Per-body gravitational potential computation (for total energy diagnostics)
- Optional collision detection using per-body radii and node-level max-radius
  bounding-volume culling (perfect for rigid-body or particle simulations)
- Validation methods (e.g., compare_with_direct) for accuracy benchmarking

The Node class is an internal building block representing either a leaf (single body)
or an internal node (subdivision). All methods are designed for repeated use in
time-stepping loops with minimal overhead.

Constants:
    G   : Gravitational constant (SI units)
    EPS : Softening parameter to prevent singularities
"""

import numpy as np


G = 6.67430e-11  # Gravitational constant in m³ kg⁻¹ s⁻² (CODATA 2018 value)
EPS = 1e-10  # Small value used to avoid division-by-zero in distance calculations


# ========================== Node ==========================
class Node:
    """
    Represents a single node in the Barnes-Hut octree.

    Each node stores spatial information (center and half-side length), aggregate
    physics quantities (total mass, center-of-mass, maximum radius when radii are
    supplied), and either a single body index (leaf node) or a list of 8 child nodes
    (internal node).

    This class is used internally by Octree and is not intended for direct
    instantiation outside the tree-building process.
    """

    def __init__(self, center: list[float] | np.ndarray, half_side: np.float64):
        """
        Initialize a new octree node.

        Parameters:
        -----------
        center : list[float] | np.ndarray
            3D coordinates of the node center (shape (3,))
        half_side : np.float64
            Half the side length of the cubic node (must be positive)

        Raises:
        -------
        ValueError
            If center is not a 3D vector or half_side is non-positive.
        """
        self.center = np.asarray(center, dtype=np.float64)

        if self.center.shape != (3,):
            raise ValueError("Node center must be a 3D vector of shape (3,)")

        if half_side <= 0:
            raise ValueError("Node half_side must be positive")

        self.half_side = half_side
        self.body_index = None  # index of the body if this is a leaf, else None
        self.children = None  # list of 8 child Node objects or None
        self.mass = 0.0  # total mass of all bodies in this subtree
        self.com = np.zeros(3)  # center-of-mass of all bodies in this subtree
        self.max_radius = (
            0.0  # maximum radius of any body in this subtree (when radii supplied)
        )


# ========================== Octree ==========================
class Octree:
    """
    Barnes-Hut octree for efficient N-body gravity and collision queries.

    This is the core spatial data structure used by GravityField. It supports:
    - Fast approximate force calculations via the Barnes-Hut criterion
    - Gravitational potential evaluation
    - Optional collision-pair detection when body radii are provided

    The tree is rebuilt from scratch for each new set of positions (typical in
    simulation time steps). Caching of the tree structure happens in the calling
    GravityField class.
    """

    def __init__(
        self,
        masses: list[float] | np.ndarray,
        pos: list[float] | np.ndarray,
        radii: list[float] | np.ndarray | None = None,
        theta: np.float64 = 0.5,
    ):
        """
        Initialize the octree with the current system state.

        Parameters:
        -----------
        masses : list[float] | np.ndarray
            List of body masses (length N)
        pos : list[float] | np.ndarray
            List of body positions (N arrays/vectors of shape (3,))
        radii : list[float] | np.ndarray | None, optional
            List of body radii for collision detection (length N). If None,
            collision methods will raise an error.
        theta : np.float64, optional
            Barnes-Hut opening angle parameter (default 0.5). Controls the
            accuracy/performance trade-off. Lower values = higher accuracy.
        """
        self.masses = masses
        self.pos = pos
        self.root = None
        self.theta = theta
        self.radii = radii
        self.collision_pairs = []

    # --------------------- Building the tree ---------------------
    def build(self) -> None:
        """
        Build the entire octree from the current set of bodies.

        Computes the bounding box of all particles, creates the root node with
        appropriate size, then inserts every body into the tree.

        Raises:
        -------
        ValueError
            If no bodies are present.
        """
        if self.pos.shape[0] == 0:
            raise ValueError(
                "Cannot build octree: positions array must contain at least one body"
            )

        x0, y0, z0 = self.pos[0]

        xmin = xmax = x0
        ymin = ymax = y0
        zmin = zmax = z0

        for p in self.pos[1:]:
            x, y, z = p

            xmin = min(xmin, x)
            xmax = max(xmax, x)

            ymin = min(ymin, y)
            ymax = max(ymax, y)

            zmin = min(zmin, z)
            zmax = max(zmax, z)

        dx = xmax - xmin
        dy = ymax - ymin
        dz = zmax - zmin

        padding = 1e-5
        half_side = max(dx, dy, dz) / 2 + padding

        cx = (xmax + xmin) / 2
        cy = (ymax + ymin) / 2
        cz = (zmax + zmin) / 2

        self.root = Node([cx, cy, cz], half_side)

        for i in range(len(self.masses)):
            self.insert(i)

    # --------------------- Inserting bodies ---------------------
    def insert(self, i) -> None:
        """
        Insert body i into the octree (public wrapper).

        Parameters:
        -----------
        i : int
            Index of the body to insert
        """
        if self.root is None:
            raise ValueError(
                "Cannot insert body: octree must be built first (call build() before insert())"
            )
        self._insert(self.root, i)

    def _insert(self, node, i) -> None:
        """
        Recursively insert body i into the subtree rooted at `node`.

        Updates mass, center-of-mass, and max_radius on every ancestor node.
        Handles leaf-to-internal promotion and subdivision automatically.
        """
        old_mass = node.mass
        new_mass = old_mass + self.masses[i]
        if old_mass == 0:
            node.com = self.pos[i].copy()
        else:
            node.com = (old_mass * node.com + self.masses[i] * self.pos[i]) / new_mass
        node.mass = new_mass

        if self.radii is not None:
            node.max_radius = max(node.max_radius, self.radii[i])

        # case 1 - empty leaf
        if node.body_index is None and node.children is None:
            node.body_index = i
            return
        # case 2 - leaf node with one body
        elif node.body_index is not None and node.children is None:
            old_body_index = node.body_index
            node.body_index = None

            node.children = self.subdivide(node)

            self._insert(
                self.choose_child(node, self.pos[old_body_index]), old_body_index
            )

            self._insert(self.choose_child(node, self.pos[i]), i)

            return
        # case 3 - internal node
        else:
            child = self.choose_child(node, self.pos[i])
            self._insert(child, i)

    # --------------------- Subdivision and child selection ---------------------
    def subdivide(self, node) -> list:
        """
        Create and return the eight child nodes for a given parent node.

        Parameters:
        -----------
        node : Node
            Parent node to subdivide

        Returns:
        --------
        list[Node]
            List of 8 new child Node objects
        """
        child_half = node.half_side / 2
        cx, cy, cz = node.center
        children = []
        for dx in (-child_half, child_half):
            for dy in (-child_half, child_half):
                for dz in (-child_half, child_half):
                    center = np.array([cx + dx, cy + dy, cz + dz], dtype=np.float64)
                    children.append(Node(center, child_half))
        return children

    def choose_child(self, node: Node, pos: np.ndarray) -> Node:
        """
        Return the appropriate child node for a given position.

        Parameters:
        -----------
        node : Node
            Parent node
        pos : np.ndarray
            Position vector (shape (3,))

        Returns:
        --------
        Node
            The child node corresponding to the octant containing `pos`
        """
        index = self.get_octant_index(node, pos)
        return node.children[index]

    def get_octant_index(self, node: Node, pos: np.ndarray) -> int:
        """
        Compute the octant index (0-7) for a position relative to a node center.

        Uses the standard binary indexing: x*4 + y*2 + z*1 where each component
        is 0 for negative and 1 for non-negative.

        Parameters:
        -----------
        node : Node
            Reference node
        pos : np.ndarray
            Position to classify

        Returns:
        --------
        int
            Octant index (0 ≤ index ≤ 7)
        """
        diff = np.subtract(pos, node.center)
        ix = 1 if diff[0] >= 0 else 0
        iy = 1 if diff[1] >= 0 else 0
        iz = 1 if diff[2] >= 0 else 0
        return ix * 4 + iy * 2 + iz

    # --------------------- Force calculation ---------------------
    def compute_acceleration(self, i: int):
        """
        Compute the gravitational acceleration on body i using the Barnes-Hut
        approximation.

        Public interface used by GravityField.__call__.

        Parameters:
        -----------
        i : int
            Index of the body

        Returns:
        --------
        np.ndarray
            Acceleration vector (shape (3,)) acting on body i
        """
        return self._compute_node_forces(self.root, i)

    def _compute_node_forces(self, node: Node, i: int):
        """
        Recursively compute the acceleration on body i due to the subtree rooted
        at `node`, applying the Barnes-Hut opening-angle criterion.
        """
        if node.mass < EPS:
            return np.zeros(3)

        a_i = np.zeros(3)

        if node.children is None and node.body_index == i:
            return np.zeros(3)

        if node.children is None and node.body_index != i:
            j = node.body_index
            disp = np.subtract(self.pos[j], self.pos[i])
            dist = max(np.linalg.norm(disp), EPS)
            f_common = G / (dist**3)
            return f_common * self.masses[j] * disp

        if node.children is not None:
            s = node.half_side * 2
            r = np.subtract(node.com, self.pos[i])
            d = np.linalg.norm(r)
            if d < EPS:
                d = EPS
            if s / d < self.theta:
                grav_coeff = G / (d**3)
                return grav_coeff * node.mass * r
            else:
                for child in node.children:
                    if child.mass == 0:
                        continue
                    a_i += self._compute_node_forces(child, i)
                return a_i

    def compare_with_direct(self, i: int) -> tuple:
        """
        Compare the tree-approximated acceleration with the exact direct summation
        for validation and accuracy benchmarking.

        Useful during development or when tuning the `theta` parameter.

        Parameters:
        -----------
        i : int
            Body index

        Returns:
        --------
        tuple[np.ndarray, np.ndarray, float]
            (tree_acc, direct_acc, absolute_difference_norm)
        """
        tree_acc = self.compute_acceleration(i)
        direct_acc = np.zeros(3)
        for j in range(len(self.masses)):
            if i == j:
                continue
            dr = np.subtract(self.pos[j], self.pos[i])
            r = max(np.linalg.norm(dr), EPS)
            direct_acc += G * self.masses[j] / r**3 * dr
        return tree_acc, direct_acc, np.linalg.norm(tree_acc - direct_acc)

    # --------------------- Collision detection ---------------------
    def point_to_node_distance(self, node: Node, position: np.ndarray) -> float:
        """
        Compute the shortest distance from a point to the surface of a node's cube.

        Used for efficient early-out culling in collision detection.

        Returns:
        --------
        float
            Distance (0 if point is inside the node)
        """
        cx, cy, cz = node.center
        h = node.half_side
        x, y, z = position

        if x < (cx - h):
            dx = (cx - h) - x
        elif x > (cx + h):
            dx = x - (cx + h)
        else:
            dx = 0

        if y < (cy - h):
            dy = (cy - h) - y
        elif y > (cy + h):
            dy = y - (cy + h)
        else:
            dy = 0

        if z < (cz - h):
            dz = (cz - h) - z
        elif z > (cz + h):
            dz = z - (cz + h)
        else:
            dz = 0

        return np.sqrt(dx**2 + dy**2 + dz**2)

    def find_collisions(self):
        """
        Find all pairs of bodies whose spheres intersect.

        Uses the octree and per-node max_radius for O(N log N) performance with
        early culling. Only works when radii were supplied at construction.

        Returns:
        --------
        list[tuple[int, int]]
            List of colliding body index pairs (i < j)
        """
        if self.radii is None:
            raise ValueError("Radii required for collision detection")

        self.collision_pairs = []
        for i in range(len(self.masses)):
            self._traverse_for_collisions(self.root, i)
        return self.collision_pairs

    def _traverse_for_collisions(self, node: Node, i: int):
        """
        Recursively traverse the subtree for collisions involving body i.
        """
        if node is None:
            return

        p_i, r_i = self.pos[i], self.radii[i]

        dist = self.point_to_node_distance(node, p_i)

        if dist > (r_i + node.max_radius):
            return

        # case one: leaf node and contain a body
        if node.children is None:
            j = node.body_index
            if j is None or i == j:
                return
            else:
                p_j, r_j = self.pos[j], self.radii[j]

                distance = np.linalg.norm(p_i - p_j)

                if distance < (r_i + r_j):
                    if i < j:
                        self.collision_pairs.append((i, j))

        # case two: internal nodes
        else:
            for child in node.children:
                self._traverse_for_collisions(child, i)

    # --------------------- Potential energy calculation ---------------------
    def compute_potential(self, i: int):
        """
        Compute the gravitational potential at body i due to all other bodies
        using the Barnes-Hut approximation.

        Public interface used by GravityField.compute_potential.

        Parameters:
        -----------
        i : int
            Body index

        Returns:
        --------
        float
            Gravitational potential at body i (excluding self-interaction)
        """
        return self._compute_node_potential(self.root, i)

    def _compute_node_potential(self, node: Node, i: int):
        """
        Recursively compute the potential at body i from the subtree rooted at
        `node`, applying the same Barnes-Hut criterion as force calculation.
        """
        # Base Case: Empty Node
        if node.mass < EPS:
            return 0.0

        phi_i = 0.0

        # Case One: Leaf Node
        if node.children is None and node.body_index is not None:
            if node.body_index == i:
                return 0.0
            else:
                j = node.body_index
                disp = np.subtract(self.pos[j], self.pos[i])
                dist = max(np.linalg.norm(disp), EPS)
                return -G * self.masses[j] / dist

        # Case Two: Internal Node
        if node.children is not None:
            s = node.half_side * 2
            r = np.subtract(node.com, self.pos[i])
            d = max(np.linalg.norm(r), EPS)

            if s / d < self.theta:
                return -G * node.mass / d
            else:
                for child in node.children:
                    if child.mass == 0:
                        continue
                    phi_i += self._compute_node_potential(child, i)
                return phi_i

File: src/neutral_physics_engine/octree.cpp
#include <array>
#include <stdexcept>
#include <memory>
#include <vector>
#include <optional>
#include <utility>
#include <algorithm>
#include <cmath>
#include <tuple>


constexpr double G = 6.67430e-11;
constexpr double EPS = 1e-10;
constexpr double eps2 = EPS * EPS;

struct vec3
{
    double x, y, z;
};
    
// mathematical operations

/* Vector Addition */
vec3 operator+(const vec3& a, const vec3& b) {
    return {
        a.x + b.x,
        a.y + b.y,
        a.z + b.z
    };
}

/* Vector Subtraction */
vec3 operator-(const vec3& a, const vec3& b) {
    return {
        a.x - b.x,
        a.y - b.y,
        a.z - b.z
    };
}

/* Vector Multiplication with scalar */
vec3 operator*(double k, const vec3& v) {
    return {k * v.x, k * v.y, k * v.z};
}

vec3 operator*(const vec3& v, double k) {
    return {v.x * k, v.y * k, v.z * k};
}

/* Vector Division with scalar */
vec3 operator/(const vec3& v, double k) {
    return {v.x / k, v.y / k, v.z / k};
}

// Dot Product of Vectors
double dot(const vec3& a, const vec3& b) {
    return a.x * b.x + a.y * b.y + a.z * b.z;
}

// Cross Product of Vectors
vec3 cross(const vec3& a, const vec3& b) {
    return {
        a.y * b.z - b.y * a.z,
        a.z * b.x - a.x * b.z,
        a.x * b.y - b.x * a.y
    };
}

// Vector Norm
double norm(const vec3& v) {
    return std::sqrt(dot(v, v));
}

// Node
struct Node
{
    vec3 center;
    double half_side;
    int body_index = -1;
    std::array<std::unique_ptr<Node>, 8> children;
    vec3 com = {0.0, 0.0, 0.0};
    double mass = 0.0;
    double max_radius = 0.0;

    // Constructor
    Node(const vec3& c, const double hs)
    : center(c), half_side(hs)
    {
        if (hs <= 0) {
            throw std::invalid_argument("Half Side should be Positive.");
        }
    }
};

// Octree
class Octree {
    private:
        const double* masses;     // Pointer to Numpy masses array [N]
        const vec3* pos;          // Pointer to Numpy N x 3 pos array (cast to vec3)
        std::unique_ptr<Node> root;
        size_t num_bodies;          // Number of bodies (size of masses and pos arrays)
        // const std::optional<std::reference_wrapper<const std::vector<double>>> radii;
        const double* radii;         // Pointer to Numpy radii array [N], optional (can be nullptr)
        double theta = 0.5;
        std::vector<std::pair<int, int>> collision_pairs;

    public:
        //Constructor
        Octree(
            const double* masses_,
            const vec3* pos_,
            const size_t num_bodies_,
            const double* radii_ = nullptr,
            const double theta_ = 0.5
        ) : masses(masses_), pos(pos_), num_bodies(num_bodies_), radii(radii_), theta(theta_)
        {
            if (num_bodies == 0 || pos == nullptr) {
                throw std::invalid_argument("Cannot build octree: positions array is empty or null.");
            }
        }

        const Node* get_root() const {
            return root.get();
        }

        // build(): building the octree
        void build() {
            // initially take first body's position as bounding box
            double xmin = pos[0].x, ymin = pos[0].y, zmin = pos[0].z;
            double xmax = pos[0].x, ymax = pos[0].y, zmax = pos[0].z;

            for (size_t i = 1; i < num_bodies; i++) {
                const auto& p = pos[i];

                if (p.x < xmin) xmin = p.x;
                if (p.y < ymin) ymin = p.y;
                if (p.z < zmin) zmin = p.z;

                if (p.x > xmax) xmax = p.x;
                if (p.y > ymax) ymax = p.y;
                if (p.z > zmax) zmax = p.z;
            }

            double half_side = std::max({xmax - xmin, ymax - ymin, zmax - zmin}) * 0.5 + 1e-5;
            
            double cx = (xmax + xmin) * 0.5, cy = (ymax + ymin) * 0.5, cz = (zmax + zmin) * 0.5;

            // create initial Node or create root
            vec3 center{cx, cy, cz};
            root = std::make_unique<Node>(center, half_side);

            // insert bodies
            for (size_t i = 0; i < num_bodies; i++) {
                insert(i);
            }
        }

        void insert(int i) {
            if (!root) throw std::invalid_argument("Call build() before insert().");

            insert_impl(root.get(), i);
        }

        void subdivide(Node* node) {
            double child_half = node->half_side * 0.5;

            int idx = 0;

            for (double dx : {-child_half, child_half}) {
                for (double dy : {-child_half, child_half}) {
                    for (double dz : {-child_half, child_half}) {
                        vec3 child_center{
                            node->center.x + dx,
                            node->center.y + dy,
                            node->center.z + dz
                        };

                        node->children[idx] = std::make_unique<Node>(child_center, child_half);
                        idx++;
                    }
                }
            }
        }

        Node* choose_child(Node* node, const vec3& pos) {
            int idx = get_octant_index(node, pos);
            return node->children[idx].get();
        }

        int get_octant_index(Node* node, const vec3& pos) {
            int ix = (pos.x >= node->center.x) ? 1 : 0;
            int iy = (pos.y >= node->center.y) ? 1 : 0;
            int iz = (pos.z >= node->center.z) ? 1 : 0;
            return ix * 4 + iy * 2 + iz;
        }

        vec3 compute_acceleration(int i) {
            vec3 acc{0.0, 0.0, 0.0};
            compute_node_forces(root.get(), i, acc);
            return acc;
        }

        std::tuple<vec3, vec3, double> compare_with_direct(int i) {
            vec3 tree_acc = compute_acceleration(i);
            vec3 direct_acc{0.0, 0.0, 0.0};

            for (size_t j = 0; j < num_bodies; j++) {
                if (static_cast<size_t>(i) == j) continue;

                double dx = pos[j].x - pos[i].x, dy = pos[j].y - pos[i].y, dz = pos[j].z - pos[i].z;
                double inv_r = 1.0 / std::sqrt(dx*dx + dy*dy + dz*dz + eps2);
                double inv_r3 = inv_r * inv_r * inv_r;
                double factor = G * masses[j] * inv_r3;

                direct_acc.x += factor * dx, direct_acc.y += factor * dy, direct_acc.z += factor * dz;
            }

            double error = norm(tree_acc - direct_acc);
            return {tree_acc, direct_acc, error};
        }

        double point_to_node_distance(Node* node, const vec3& pos) {
            double dx, dy, dz;
            if (pos.x < (node->center.x - node->half_side)) dx = (node->center.x - node->half_side) - pos.x;
            else if (pos.x > (node->center.x + node->half_side)) dx = pos.x - (node->center.x + node->half_side);
            else dx = 0.0;

            if (pos.y < (node->center.y - node->half_side)) dy = (node->center.y - node->half_side) - pos.y;
            else if (pos.y > (node->center.y + node->half_side)) dy = pos.y - (node->center.y + node->half_side);
            else dy = 0.0;

            if (pos.z < (node->center.z - node->half_side)) dz = (node->center.z - node->half_side) - pos.z;
            else if (pos.z > (node->center.z + node->half_side)) dz = pos.z - (node->center.z + node->half_side);
            else dz = 0.0;

            return std::sqrt(dx*dx + dy*dy + dz*dz);
        } 

        std::vector<std::pair<int, int>> find_collisions() {
            if (!root) throw std::invalid_argument("Build octree before collision detection.");

            if (!radii) throw std::invalid_argument("Radii are required for collision detection.");

            collision_pairs.clear();

            for (size_t i = 0; i < num_bodies; i++) {
                traverse_for_collisions(root.get(), i);
            }

            return collision_pairs;
        }

        double compute_potential(int i) {
            if (!root) throw std::invalid_argument("call build() first.");
            return compute_node_potential(root.get(), i);
        }

    private:
        void insert_impl(Node* node, int i) {
            double old_mass = node->mass;
            double new_mass = old_mass + masses[i];
            const vec3& r = pos[i];

            if (old_mass == 0) {
                node->com = r;
            } else {
                node->com = (node->com * old_mass + r * masses[i]) / new_mass;
            }
            node->mass = new_mass;

            if (radii) {
                node->max_radius = std::max(node->max_radius, radii[i]);
            }

            // case one: Empty Leaf
            if (node->body_index == -1 && node->children[0] == nullptr) {
                node->body_index = i;
                return;
            }

            // case two: Leaf with One Body
            else if (node->body_index != -1 && node->children[0] == nullptr)
            {
                int old_body_index = node->body_index;

                node->body_index = -1;

                subdivide(node);

                insert_impl(choose_child(node, pos[old_body_index]), old_body_index);

                insert_impl(choose_child(node, pos[i]), i);

                return;
            }

            // case three: Internal Node
            else {
                insert_impl(choose_child(node, pos[i]), i);
            }
            
        }

        void compute_node_forces(Node* node, int i, vec3& acc) {
            if (node->mass < EPS) return;

            if (node->children[0] == nullptr && node->body_index == i) return;

            if (node->children[0] == nullptr && node->body_index != i) {
                int j = node->body_index;
                double dx = pos[j].x - pos[i].x, dy = pos[j].y - pos[i].y, dz = pos[j].z - pos[i].z;
                double inv_dist = 1.0 / std::sqrt(dx*dx + dy*dy + dz*dz + eps2);
                double inv_dist3 = inv_dist * inv_dist * inv_dist;
                double f_common = G * masses[j] * inv_dist3;

                acc.x += f_common * dx, acc.y += f_common * dy, acc.z += f_common * dz;
                return;
            }

            double s = node->half_side * 2.0;
            double rx = node->com.x - pos[i].x, ry = node->com.y - pos[i].y, rz = node->com.z - pos[i].z;
            double d = std::sqrt(rx*rx + ry*ry + rz*rz + eps2);
            if (s / d < theta) {
                double inv_d = 1.0 / d;
                double inv_d3 = inv_d * inv_d * inv_d;
                double factor = G * node->mass * inv_d3;

                acc.x += factor * rx, acc.y += factor * ry, acc.z += factor * rz;
                return;
            }

            else {
                for (const auto& child : node->children) {
                    if (!child || child->mass == 0.0) continue;

                    compute_node_forces(child.get(), i, acc);
                }
            }
        }

        void traverse_for_collisions(Node* node, int i) {
            if (!node) return;

            const vec3& p_i = pos[i];
            double r_i = radii[i];

            double dist = point_to_node_distance(node, p_i);

            if (dist > (r_i + node->max_radius)) return;

            // case one: Leat node contain one body
            if (node->children[0] == nullptr) {
                int j = node->body_index;
                
                if (j == -1 || i == j) return;
                else {
                    double dx = p_i.x - pos[j].x, dy = p_i.y - pos[j].y, dz = p_i.z - pos[j].z;
                    
                    double distance = std::sqrt(dx*dx + dy*dy + dz*dz);
                    double radius_sum = r_i + radii[j];

                    if (distance < radius_sum) {
                        if (i < j) collision_pairs.emplace_back(i, j);
                    }
                }
            }

            // case two: internal nodes
            else {
                for (const auto& child : node->children) {
                    traverse_for_collisions(child.get(), i);
                }
            }
        }

        double compute_node_potential(Node* node, int i) {
            // base case: Empty Node
            if (node->mass == 0.0) return 0.0;

            double phi_i = 0.0;

            // case one: Leaf Node
            if (node->children[0] == nullptr && node->body_index != -1) {
                if (node->body_index == i) return 0.0;

                else {
                    int j = node->body_index;
                    double dx = pos[j].x - pos[i].x, dy = pos[j].y - pos[i].y, dz = pos[j].z - pos[i].z;
                    double inv_dist = 1.0 / std::sqrt(dx*dx + dy*dy + dz*dz + eps2);
                    return -G * masses[j] * inv_dist;
                }
            }

            // case two: Internal Node
            double s = node->half_side * 2.0;
            double rx = node->com.x - pos[i].x, ry = node->com.y - pos[i].y, rz = node->com.z - pos[i].z;
            double d = std::sqrt(rx*rx + ry*ry + rz*rz + eps2);

            if (s / d < theta) {
                double inv_d = 1.0 / d;
                return -G * node->mass * inv_d;
            }

            else {
                for (const auto& child : node->children) {
                    if (!child || child->mass == 0.0) continue;
                    
                    phi_i += compute_node_potential(child.get(), i);
                }
                return phi_i;
            }
        }

};